src/prepared_statement.c

```diff
@@ -0,0 +1,572 @@
+#include "bouncer.h"
+
+#include <usual/crypto/csrandom.h>
+#include <usual/hashtab-impl.h>
+#include <usual/slab.h>
+
+static PgPreparedStatement* prepared_statements;
+static queryid_t next_unique_query_id;
+static struct Slab* server_ps_cache;
+
+/*
+ * Like uthash HASH_DELETE, but doesn't remove item from hash table just
+ * unlink element from the doubly-linked-list.
+ */
+#define HASH_UNLINK(hh, head, delptr)									\
+	do {																\
+		struct UT_hash_handle *_hd_hh_del = &(delptr)->hh;				\
+		if (_hd_hh_del->prev != NULL) {									\
+			HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->prev)->next = _hd_hh_del->next; \
+		} else {														\
+			DECLTYPE_ASSIGN(head, _hd_hh_del->next);					\
+		}																\
+		if (_hd_hh_del->next != NULL) {									\
+			HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->next)->prev = _hd_hh_del->prev; \
+		} else {														\
+			_hd_hh_del->tbl->tail = HH_FROM_ELMT(_hd_hh_del->tbl, _hd_hh_del->prev); \
+		}																\
+	} while (0)
+
+static PgPreparedStatement*
+create_prepared_statement(PgParsePacket *pkt)
+{
+	size_t query_len = strlen(pkt->query) + 1;
+	size_t num_parameters = pkt->num_parameters * 4;
+	PgPreparedStatement *ps = (PgPreparedStatement *)malloc(sizeof(PgPreparedStatement) + query_len + num_parameters);
+	if (ps != NULL)
+	{
+		ps->query_len = query_len;
+		ps->param_len = num_parameters;
+		ps->query_id = next_unique_query_id;
+		ps->use_count = 0;
+		memcpy(ps->query_with_params, pkt->query, query_len);
+		memcpy(&ps->query_with_params[query_len], pkt->parameter_types_bytes, num_parameters);
+	}
+	return ps;
+}
+
+
+void
+ps_init(void)
+{
+	server_ps_cache = slab_create("server_ps_cache", sizeof(PgServerPreparedStatement), 0, NULL, USUAL_ALLOC);
+}
+
+void
+ps_cleanup(void)
+{
+	slab_destroy(server_ps_cache);
+	server_ps_cache = NULL;
+}
+
+
+static PgClientPreparedStatement*
+create_client_prepared_statement(char const* name, PgPreparedStatement* ps)
+{
+	size_t name_len = strlen(name) + 1;
+	PgClientPreparedStatement *client_ps = (PgClientPreparedStatement*)malloc(sizeof(PgClientPreparedStatement) + name_len);
+	if (client_ps != NULL)
+	{
+		memcpy(client_ps->stmt_name, name, name_len);
+		client_ps->ps = ps;
+		ps->use_count += 1;
+	}
+	return client_ps;
+}
+
+static PgServerPreparedStatement*
+create_server_prepared_statement(PgPreparedStatement* ps)
+{
+	PgServerPreparedStatement *server_ps = (PgServerPreparedStatement*)slab_alloc(server_ps_cache);
+	if (server_ps != NULL)
+	{
+		server_ps->ps = ps;
+		server_ps->query_id = ps->query_id;
+		ps->use_count += 1;
+	}
+	return server_ps;
+}
+
+static PgPreparedStatement*
+get_prepared_statement(PgParsePacket *pkt, bool* found)
+{
+	PgPreparedStatement *ps = create_prepared_statement(pkt);
+	PgPreparedStatement *old_ps = NULL;
+	if (ps != NULL)
+	{
+		HASH_FIND(hh, prepared_statements, ps->query_with_params, ps->query_len + ps->param_len, old_ps);
+		if (old_ps != NULL)
+		{
+			free(ps);
```

Comment:
It would be good not to allocate + free here. And instead search the hashmap using the data from the PgParsePacket.

Reply:
Yeah, I also will be glad to avoid memory allocation here.
But we need to construct key including query parameters. Them are located in different parts of parse packets and Certainly we can redefine  HASH_KEYCMP and HASH_KEYCMP to deal with scattered hash key.
But it will be less efficient and I am not sure that it will work at all, because actually perpared_statement.c is using 3 kind of hashes (by int key, by string key and by query text + parameters).

Reply:
I pushed a change now that removes this allocation. The way I did it is by changing the hash key to be all the raw bytes of the Parse packet starting at the first byte of the query string. This does make the hash key 3 bytes longer (1 for the null byte and 2 for the number of parameters), but that seems a small price to pay to avoid this allocation. And I think it made the code a bit simpler too.